test/NCI.OCPL.Api.Glossary.Tests/Tests/Controllers/TermsControllerTest.GetByName.cs

```diff
@@ -133,5 +135,237 @@ public async void GetByNameTerms()
+                termQSvc => termQSvc.GetByName(
+                    It.Is<String>(dictionary => dictionary == "Genetics"),
+                    It.Is<AudienceType>(audience => audience == AudienceType.HealthProfessional),
+                    It.Is<string>(language => language == "en"),
+                    It.Is<string>(prettyUrlName => prettyUrlName == "s-phase-fraction")
+                )
+            )
+            .Callback(() => Assert.Equal(4, callOrder++))
+            .Throws(new APIErrorException(200, "Empty response when searching for dictionary 'Genetics', audience 'HealthProfessional', language 'en', pretty URL name 's-phase-fraction'."));
+
+            termQueryService.Setup(
+                termQSvc => termQSvc.GetByName(
+                    It.Is<String>(dictionary => dictionary == "Cancer.gov"),
+                    It.Is<AudienceType>(audience => audience == AudienceType.Patient),
+                    It.Is<string>(language => language == "en"),
+                    It.Is<string>(prettyUrlName => prettyUrlName == "s-phase-fraction")
+                )
+            )
+            .Returns(Task.FromResult(glossaryTerm));
+
+            TermsController controller = new TermsController(NullLogger<TermsController>.Instance, termQueryService.Object);
+            GlossaryTerm gsTerm = await controller.GetByName("NotSet", AudienceType.Patient, "en", "s-phase-fraction", true);
+            string actualJsonValue = JsonConvert.SerializeObject(gsTerm);
+            string expectedJsonValue = File.ReadAllText(TestingTools.GetPathToTestFile("TermsControllerData/TestData_GetWithFallback_TermsPatient.json"));
+
+            // Verify that the service layer is called correctly with the fallback logic:
+            // 1) Empty dictionary, Patient
+            termQueryService.Verify(
+                svc => svc.GetByName("NotSet", AudienceType.Patient, "en", "s-phase-fraction"),
+                Times.Once
+            );
+            // 2) Empty dictionary, HealthProfessional
+            termQueryService.Verify(
+                svc => svc.GetByName("NotSet", AudienceType.HealthProfessional, "en", "s-phase-fraction"),
+                Times.Once
+            );
+            // 3) Genetics, Patient
+            termQueryService.Verify(
+                svc => svc.GetByName("Genetics", AudienceType.Patient, "en", "s-phase-fraction"),
+                Times.Once
+            );
+            // 4) Genetics, HealthProfessional
+            termQueryService.Verify(
+                svc => svc.GetByName("Genetics", AudienceType.HealthProfessional, "en", "s-phase-fraction"),
+                Times.Once
+            );
+            // 5) Cancer.gov, Patient
+            termQueryService.Verify(
+                svc => svc.GetByName("Cancer.gov", AudienceType.Patient, "en", "s-phase-fraction"),
+                Times.Once
+            );
+
+            Assert.Equal(expectedJsonValue, actualJsonValue);
+        }
     }
 }
```

Comment:
Needs a fallback test for the case where the dictionary is "NotSet".

Reply:
The fallback tests are only for Patient vs HealthProfessional fallback logic - not based on which dictionary we come in with. The HealthProfessional test starts with Cancer.gov/HP, and the Patient test starts with NotSet/Patient. The tests perform the longest loop around the fallback logic options to make sure that they work correctly.

Reply:
You raise several good points.  I was going on the basis that these test were pulling double duty, testing both audience and dictionary. The choice of starting place varies with which dictionary is selected, that should be tested too, 

Audience test can be combined with dictionary however, otherwise the number of tests grows hugely.